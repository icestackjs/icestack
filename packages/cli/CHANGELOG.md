# @icestack/cli

## 1.2.0

### Minor Changes

- [`7706a6a`](https://github.com/sonofmagic/icestack/commit/7706a6aa6ef786308612e9680104fcc485cdd6d0) Thanks [@sonofmagic](https://github.com/sonofmagic)! - add build cache and postcss plugin

### Patch Changes

- Updated dependencies [[`7706a6a`](https://github.com/sonofmagic/icestack/commit/7706a6aa6ef786308612e9680104fcc485cdd6d0)]:
  - @icestack/config@1.2.0
  - @icestack/core@1.2.0
  - @icestack/logger@1.2.0
  - @icestack/shared@1.2.0
