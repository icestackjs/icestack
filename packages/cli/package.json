{
  "name": "@icestack/cli",
  "description": "The cli of @icestack/ui",
  "version": "1.3.4",
  "scripts": {
    "build": "unbuild",
    "test": "vitest run",
    "test:dev": "vitest",
    "release": "pnpm publish",
    "sync": "cnpm sync @icestack/cli"
  },
  "keywords": [
    "cli",
    "component",
    "generate",
    "generator",
    "tailwindcss",
    "unocss",
    "icestack",
    "ui"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "dependencies": {
    "@icestack/config": "workspace:^",
    "@icestack/core": "workspace:^",
    "@icestack/logger": "workspace:^",
    "@icestack/shared": "workspace:^",
    "chokidar": "^3.5.3",
    "commander": "^12.0.0",
    "dedent": "^1.5.1",
    "del": "^7.1.0"
  },
  "author": "SonOfMagic <qq1324318532@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/sonofmagic/icestack"
  },
  "homepage": "https://ui.icebreaker.top",
  "repository": {
    "type": "git",
    "url": "https://github.com/sonofmagic/icestack"
  }
}
