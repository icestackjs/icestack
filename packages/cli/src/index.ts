import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { loadSync } from '@icestack/config'
import { createContext } from '@icestack/core'
import { logger } from '@icestack/logger'
import { getModuleDependencies, touch } from '@icestack/shared'
import chokidar from 'chokidar'
import { Command } from 'commander'
import { createJsConfig } from './config'

function touchTwConfig(filepath: string) {
  const dirname = path.dirname(filepath)
  const configPath = path.resolve(dirname, 'tailwind.config.js')
  if (fs.existsSync(configPath)) {
    touch(configPath)
  }
}

export const cli = new Command()

async function letUsBuild(options: { clean?: boolean, configFile?: string } = {}) {
  const { clean, configFile } = options
  if (configFile) {
    logger.success(`load config from ${configFile}`)
  }
  const o = await loadSync({
    configFile,
  })
  if (!o) {
    logger.error('load options fail!')
    return
  }
  const { config, filepath } = o
  if (config) {
    if (!config.outdir) {
      logger.error('outdir option must be passed!')
      return
    }
    if (clean || config.clean) {
      const { deleteAsync } = await import('del')
      await deleteAsync(config.outdir)
      logger.success(`del outdir: ${config.outdir} successfully!`)
    }

    const ctx = createContext(filepath)
    await ctx.build()
    logger.success('build successfully!')
  }
  return o
}
const defaultOutdir = 'my-ui'

cli
  .command('init')
  .description('init config')
  .option('--only-config', 'output onlt config')
  .option('--format <format>', 'config file format')
  .option('--outdir <outdir>')
  .option('--mode <mode>')
  .action(async (options) => {
    const cwd = process.cwd()
    const { onlyConfig, outdir = defaultOutdir, format = 'cjs', mode = 'none' } = options
    const { data, filename } = createJsConfig({
      outdir,
      format,
      mode,
    })
    const p = path.resolve(cwd, filename)
    fs.writeFileSync(p, data, 'utf8') // ts ? tsT : jsT)
    logger.success(`init ${filename} successfully!`)
    if (onlyConfig) {
      return
    }
    const gitignoreFilePath = path.resolve(cwd, '.gitignore')
    fs.existsSync(gitignoreFilePath) ? fs.appendFileSync(gitignoreFilePath, `\n${outdir}\n`, 'utf8') : fs.writeFileSync(gitignoreFilePath, `${outdir}\n`, 'utf8')

    await letUsBuild()
  })

cli
  .command('build')
  .description('code generate')
  .option('--clean')
  .option('-c, --config <path>', 'config path')
  .action(async (options) => {
    const { clean, config } = options
    const opts = { clean, configFile: config }
    const res = await letUsBuild(opts)
    res && touchTwConfig(res.filepath)
  })

cli
  .command('watch')
  .description('watch config file change and build lib')
  .option('-c, --config <path>', 'config path')
  .action(async (options) => {
    const { clean, config } = options
    const opts = { clean, configFile: config }
    const { configFile } = opts
    const cwd = process.cwd()
    const o = await loadSync({
      cwd,
      configFile,
    })
    if (!o) {
      logger.error('load options fail!')
      return
    }
    const { filepath } = o
    if (configFile) {
      logger.success(`load config from ${configFile}`)
    }
    // path.resolve(cwd, 'icestack.config.{js,ts,cjs}')
    chokidar
      .watch([...getModuleDependencies(filepath)])
      .on('change', async () => {
        const res = await letUsBuild(opts)
        res && touchTwConfig(res.filepath)
        logger.success('some changes happened')
      })
      .on('ready', async () => {
        const res = await letUsBuild(opts)
        res && touchTwConfig(res.filepath)
        logger.success('start watching config......')
      })
  })
