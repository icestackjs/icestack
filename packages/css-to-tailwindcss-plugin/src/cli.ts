import path from 'node:path'
import fs from 'node:fs'
import createCac from 'cac'
import { createContext } from './core'
import { ensureDir, resolvePath } from './utils'
import { LayerEnumType, version } from '@/constants'
const cli = createCac()

const defaultCwd = process.cwd()

cli
  .command('build [...files]', 'Build files')
  .option('--out <dir>', 'Output directory')
  .option('--cwd <cwd>', 'Current working directory')
  .option('--resolved, --tailwindcssResolved', 'If Resolved tailwindcss ')
  .option('-c, --config, --tailwindcssConfig <config>', 'Tailwindcss config path')
  .action(async (files: string[], options: { out: string; cwd: string; outSideLayerCss: LayerEnumType; tailwindcssResolved: boolean; tailwindcssConfig: string }) => {
    const { cwd, out, outSideLayerCss, tailwindcssResolved, tailwindcssConfig } = options
    for (const file of files) {
      const entry = resolvePath(file, resolvePath(cwd, defaultCwd))
      if (!fs.existsSync(entry)) {
        console.log(`${file} isn't existed! skipped`)
        continue
      }
      const ctx = createContext({
        outSideLayerCss,
        tailwindcssResolved,
        tailwindcssConfig
      })
      await ctx.process(entry)
      const code = ctx.generate()
      const filename = path.basename(entry, path.extname(entry))
      const outDir = out ? resolvePath(out, cwd) : path.dirname(entry)
      ensureDir(outDir)
      const target = path.resolve(outDir, filename + '.js')
      fs.writeFileSync(target, code, 'utf8')
      console.log(`build successfully! file: ${target}`)
    }
  })

cli.help()

cli.version(version)

cli.parse()
