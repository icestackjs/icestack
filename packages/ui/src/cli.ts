import fs from 'node:fs/promises'
import path from 'node:path'
import createCli from 'cac'
import { loadConfig } from 'c12'
import type { CodegenOptions, DeepPartial } from './types'
import { buildAll } from './generate'
import { getCodegenOptions } from './options'
import { logger } from '@/log'
const cli = createCli()

export async function load(cwd?: string) {
  const { config } = await loadConfig<DeepPartial<CodegenOptions>>({
    name: 'icestack',
    cwd,
    defaultConfig: getCodegenOptions()
  })
  return config
}

const tsT = `import { defineConfig } from '@icestack/ui'\n\nexport default defineConfig({
  outdir: './my-ui'
})\n`
const jsT = `/**\n * @type {import('@icestack/ui').Config}\n */\nconst config = {
  outdir: './my-ui'
}\n\nmodule.exports = config\n`
cli
  .command('init', 'init config')
  .option('--ts', 'typescript config')
  .action(async ({ ts }) => {
    const f = ts ? 'icestack.config.ts' : 'icestack.config.js'
    const p = path.resolve(process.cwd(), f)
    await fs.writeFile(p, ts ? tsT : jsT)
    logger.success(`init ${f} successfully!`)
  })

cli.command('build', 'code generate').action(async () => {
  const config = await load()
  if (config) {
    if (!config.outdir) {
      logger.error('outdir option must be passed!')
      return
    }
    const cfg = getCodegenOptions(config, true)
    await buildAll(cfg)
    logger.success('build successfully!')
  }
})

cli.help()
cli.version('0.0.0')
cli.parse()
