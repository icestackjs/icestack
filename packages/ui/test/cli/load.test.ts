import { load } from '@/cli'

describe('load config', () => {
  it('load config', async () => {
    const config = await load(__dirname)
    expect(config).toMatchSnapshot()
  })
})
