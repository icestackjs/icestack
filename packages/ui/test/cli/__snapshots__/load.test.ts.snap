// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`load config > load cjs config 1`] = `
{
  "base": {
    "themes": {
      "dark": {
        "selector": ".dark",
      },
      "light": {
        "selector": "page",
      },
    },
  },
  "log": false,
}
`;

exports[`load config > load js config 1`] = `
{
  "base": {
    "themes": {
      "dark": {
        "selector": ".dark",
      },
      "light": {
        "selector": "page",
      },
    },
  },
  "log": false,
}
`;

exports[`load config > load mjs config 1`] = `
{
  "base": {
    "themes": {
      "dark": {
        "selector": ".dark",
      },
      "light": {
        "selector": "page",
      },
    },
  },
  "log": false,
}
`;

exports[`load config > load ts config 1`] = `
{
  "base": {
    "themes": {
      "dark": {
        "selector": ".dark",
      },
      "light": {
        "selector": "page",
      },
    },
  },
  "log": false,
}
`;
